Move logging state to per-CPU flags.

qwattash opened the pull request:
Instead of using the global CPU_LOG_INSTR and CPU_LOG_USER_ONLY flags in `qemu_loglevel`, the instruction logging level is handled per-CPU.
The switching is performed asynchronously in the CPU exclusive context, as we have to get there to flush the CPU TCG buffer anyway.

qwattash committed:
mips: Fix logging of EntryLo1.

Use the correct CP0 register index to find the register name.

qwattash committed:
Refactor instruction logging pause/resume.

Move to using per-CPU log levels. Each CPU has associated a separate log
strategy (currently ALL, NONE, USER-ONLY), and a pause/resume flag.
The CPU_LOG_INSTR level is the global enable for instruction logging,
this is set at startup and never (normally) unset, we still need the flag to
keep the trace file open.

The pause/resume switching logic is moved to a callback that is run
within the CPU exclusive context at the same time of flushing the TCG
buffer. This makes the switching logic cleaner and the synchronization
is easier to handle. We also are guaranteed to be called after the side-effects
for the current instructions are completely handled by the target.

ctsrd-jenkins commented:
Changes to base branch detected, but this pull request has already been tested.
 To re-test against latest HEAD trigger a build manually or add the 'ALWAYS-JENKINS' label.

arichardson commented:
> Changes to base branch detected, but this pull request has already been tested.
> To re-test against latest HEAD trigger a build manually or add the 'ALWAYS-JENKINS' label.

Guess I need to fix this incorrect message

qwattash committed:
Fix build with instruction logging disabled.

Disable instruction logging by default, as we do not support it for all targets.
Instead enable it explicitly at configuration time.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -193,11 +197,17 @@ static trace_fmt_hooks_t *trace_format = NULL;
-    ((old & mask) != (requested & mask))
+/* Helper to check whether a tracing mode bit was switched on */
+#define loglevel_enabled(bitmask, old, requested)               \
+    ((old & bitmaks) == 0 && (requested & bitmask) != 0)
```
```suggestion
    ((old & bitmask) == 0 && (requested & bitmask) != 0)
```

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -193,11 +197,17 @@ static trace_fmt_hooks_t *trace_format = NULL;
+
+/* Helper to check whether a tracing mode bit was switched off  */
+#define loglevel_disabled(bitmask, old, requested)              \
+    ((old & bitmaks) == 0 && (requested & bitmask) != 0)
```
```suggestion
    ((old & bitmask) == 0 && (requested & bitmask) != 0)
```

arichardson replied:
I guess these are not used? Otherwise this would have caused errors?

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -397,16 +407,16 @@ static void emit_text_start(CPUArchState *env, target_ulong pc)
-        qemu_log("Requested instruction logging @ " TARGET_FMT_lx
-                 " ASID %u\n", pc, cpu_get_asid(env));
+    switch (cpulog->loglevel) {
+    case QEMU_LOG_INSTR_LOGLVL_USER:
```
QEMU_LOG_INSTR_LOGLVL_USER is so long already, I think expanding it to LEVEL might be better.

qwattash replied:
Good point. I should probably shorten them but I can't seem to find a satisfactory prefix given that I have been using the `qemu_log_instr_*` for everything exported by this module.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -416,12 +426,16 @@ static void emit_text_stop(CPUArchState *env, target_ulong pc)
-    else
-        qemu_log("Disabled instruction logging @ " TARGET_FMT_lx
-                 " ASID %u\n", pc, cpu_get_asid(env));
+    switch (cpulog->loglevel) {
```
Here and above: I'm not sure using a switch for only one case is ideal. Unless you are planning to add more in the future?

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -581,37 +605,90 @@ static void do_instr_commit(CPUArchState *env)
 
-    cpulog->user_mode_tracing = false;
-    cpulog->instr_info = iinfo;
+    /* Decide whether if we have to pause/resume logging */
```
```suggestion
    /* Decide whetherwe have to pause/resume logging */
```

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -581,37 +605,90 @@ static void do_instr_commit(CPUArchState *env)
+    cpulog->loglevel_active = next_level_active;
+
+    /* Check if this was a no-op */
+    if (next_level == prev_level && prev_level_active == next_level_active)
+        return;
+    else {
```
```suggestion
    if (next_level == prev_level && prev_level_active == next_level_active) {
        return;
    } else {
```

arichardson replied:
Or possibly better: drop the else since you have an early return.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -965,36 +965,72 @@ void qemu_log_instr_extra(CPUArchState *env, const char *msg, ...)
+    cpu_loglevel_switch(env, QEMU_LOG_INSTR_LOGLVL_USER);
 }
 
+/* Start logging on the current CPU */
```
```suggestion
/* Stop logging on the current CPU */
```

arichardson commented on include/qemu/log_instr.h:
```diff
@@ -67,6 +68,18 @@ typedef enum {
+ */
+typedef enum {
+    /* No logging for this CPU */
+    QEMU_LOG_INSTR_LOGLVL_NONE = 0,
```
I would spell these out as `QEMU_LOG_INSTR_LOGLEVEL_*`

arichardson reviewed (approved):
I haven't looked at the logic in detail so mostly cosmestic suggestions. Otherwise LGTM If it works for you. I think having this state per-CPU is a nice improvement.

Can we drop CPU_LOG_USER_ONLY now?

qwattash committed:
Cleanup instruction logging subsystem.

Remove CPU_LOG_USER_ONLY as it is now unused.
Remove unused macros.
Improve naming of per-CPU log levels.

qwattash committed:
Fix bug in stopping instruction tracing from qemu monitor.

Clearing the CPU_LOG_INSTR flag immediately results in empty instructions
being committed until the end of the current TB because the we do not
check for the flag in commit.
We now delay stopping everything until the end of the TB.
